### PR TITLE
Implement grpc.Future interface in SingleThreadedRendezvous

### PR DESCRIPTION
**Pull request by gnossen**
This adds partial support for the `grpc.Future` interface to the experimental `_SingleThreadedRendezvous` class. It implements the interface in every case that does not require blocking. In cases where a call would otherwise block, a `UsageError` is raised instead. This is sufficient to call `add_done_callback`, which is useful for implementing interceptors.

### CODE REVIEW

**Comment by gnossen** [comment]
CC @benrkarl

File: src/python/grpcio/grpc/_channel.py [hunk 1]
```diff
@@ -442,15 +442,103 @@ def __del__(self):
...
+    def running(self):
+        with self._state.condition:
+            return self._state.code is None
+
+    def done(self):
+        with self._state.condition:
+            return self._state.code is not None
+
+    def result(self, timeout=None):
+        """Returns the result of the computation or raises its exception.
```
Comment (lidizheng) [inline src/python/grpcio/grpc/_channel.py:475]: nit: Should we add a short comment about why timeout is not feasible for one-threaded future? Same for other comments.

**Review by lidizheng — APPROVED** [review]
LGTM!

**Comment by gnossen** [comment]
Looks like one of the artifact builds ended up scheduled on a 32 bit machine somehow:

```
INFO: Running bazel wrapper (see //tools/bazel for details), bazel version 2.2.0 will be used instead of system-wide bazel installation.
Unsupported architecture: Linux i686
Invoking "bazel query" to gather the protobuf dependencies.
Running "bazel query //:protoc_lib"
Got non-fatal error:
```

**Pull request merged**